pkg/registry/distribution/client/repo.go

```diff
@@ -0,0 +1,282 @@
+	if err != nil {
+		return err
+	}
+
+	if !exist {
+		log.Warnf("repo: %s:%s manifests not found.", repoName, tag)
+		return nil
+	}
+
+	if err := r.DeleteManifest(digest); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ListTag ...
+func (r *Repository) ListTag() ([]string, error) {
+	tags := []string{}
+	req, err := http.NewRequest("GET", buildTagListURL(r.Endpoint.String(), r.Name), nil)
+	if err != nil {
+		return tags, err
+	}
+	err = r.withAuthInfo(req)
+	if err != nil {
+		return tags, err
+	}
+
+	resp, err := r.client.Do(req)
+	if err != nil {
+		return tags, parseError(err)
+	}
+
+	defer resp.Body.Close()
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return tags, err
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		tagsResp := struct {
+			Tags []string `json:"tags"`
+		}{}
+
+		if err := json.Unmarshal(b, &tagsResp); err != nil {
+			return tags, err
+		}
+		sort.Strings(tags)
+		tags = tagsResp.Tags
+
+		return tags, nil
+	} else if resp.StatusCode == http.StatusNotFound {
+		return tags, nil
+	}
+
+	return tags, &Error{
+		Code:    resp.StatusCode,
+		Message: string(b),
+	}
+
+}
+
+// ManifestExist ...
+func (r *Repository) ManifestExist(reference string) (digest string, exist bool, err error) {
+	req, err := http.NewRequest("HEAD", buildManifestURL(r.Endpoint.String(), r.Name, reference), nil)
+	if err != nil {
+		return
+	}
+	err = r.withAuthInfo(req)
+	if err != nil {
+		return
+	}
+
+	for _, mediaType := range ManifestAccepts {
+		req.Header.Add("Accept", mediaType)
+	}
+
+	resp, err := r.client.Do(req)
+	if err != nil {
+		err = parseError(err)
+		return
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		exist = true
+		digest = resp.Header.Get("Docker-Content-Digest")
+		return
+	}
+
+	if resp.StatusCode == http.StatusNotFound {
+		return
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return
+	}
+
+	err = &Error{
+		Code:    resp.StatusCode,
+		Message: string(b),
+	}
+	return
+}
+
+// DeleteManifest ...
+func (r *Repository) DeleteManifest(digest string) error {
+	req, err := http.NewRequest("DELETE", buildManifestURL(r.Endpoint.String(), r.Name, digest), nil)
+	if err != nil {
+		return err
+	}
+	err = r.withAuthInfo(req)
+	if err != nil {
+		return err
+	}
+
+	resp, err := r.client.Do(req)
+	if err != nil {
+		return parseError(err)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusAccepted {
+		return nil
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return &Error{
+		Code:    resp.StatusCode,
+		Message: string(b),
+	}
+}
+
+func (r *Repository) withAuthInfo(req *http.Request) error {
+	access := []*token.ResourceActions{
+		{
+			Type:    "repository",
+			Actions: []string{"*", "pull"},
+			// to make token be available, should rename repo name with tenantID
+			Name: fmt.Sprintf("%s-%s", r.TenantID, r.Name),
+		},
+	}
+	token, err := auth.MakeToken("admin", access, 24, r.privateKey)
```

Comment:
please get user name from http requet rather than hard code to `admin`

Reply:
done